Homework Sprint #15.

nim0y opened the pull request:
Подключил:
ModelView(не везде , моральных сил уже нет)
Koin
Попытался в корутины ничего путного не вышло.
Для будущих спринтов заведу отдельную ветку.
Анна,
Не передать словами испытываемое мной чувство благодарности к тебе (надеюсь можно на ты).
Не только за огромный обьём проделанной тобой работы.
Но и за то что с пониманием отнеслась к моей "слетевшей кукухе".
Постарался подчистить то что ты описывала в ревью и комментариях. Не всё получилось, но я продолжаю работу во многом благодаря тебе.
Еще раз спасибо!

nim0y committed:
Homework Sprint #15.
Подключил:
ModelView(не везде , моральных сил уже нет)
Koin
Попытался в корутины ничего путного не вышло.
Переписывал всё под чистую : 8 раз, скорее всего породил сикстиллион новых багов.
Не понимаю как убрать баг Switch на тёмную тему.(спрошу у  наставников обязательно)
"Любой, имеющий в доме ружьё, приравнивается к Курту Кобейну
 Любой, умеющий читать между строк, обречён иметь в доме ружьё"
 Просто устал...

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/data/impl/player/PlayerRepositoryImpl.kt:
```diff
@@ -1,13 +1,13 @@
-class PlayerImpl : Player {
+class PlayerRepositoryImpl : PlayerRepository {
 
     private var mediaPlayer = MediaPlayer()
```
плеер тоже должен быть внедрен через koin, это будет проверяться в 17 спринте

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/data/impl/search/HistoryRepositoryImpl.kt:
```diff
@@ -0,0 +1,37 @@
+
+class HistoryRepositoryImpl(context: Context, private val gson: Gson) : HistoryRepository {
+    private val sharedPreferences: SharedPreferences =
+        context.getSharedPreferences(PREFER_SEARCH, Application.MODE_PRIVATE)
```
И это тоже к 17 спринту надо будет через коин передавать уже готовым

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/data/impl/search/HistoryRepositoryImpl.kt:
```diff
@@ -0,0 +1,37 @@
+    override fun read(): List<Track> {
+        val json = sharedPreferences.getString(PREF_KEY, null) ?: return emptyList()
+        return Gson().fromJson(json, Array<Track>::class.java).toList()
+        //Gson().fromJson(json, Array<Track>::class.java).toList()
```
Удалить можно

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/data/dto/Constants.kt:
```diff
@@ -0,0 +1,14 @@
+package com.example.playlistmaker.data.dto
+
+object Constants {
+    const val PREFER_SEARCH = "shared_pref"
```
Константы можно еще писать сразу на уровне класса без дополнительных объектов/классов

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/data/dto/Constants.kt:
```diff
@@ -0,0 +1,14 @@
+    const val MAX_TEMP_LIST_SIZE = 10
+    const val ITUNES_URL = "https://itunes.apple.com"
+    const val SEARCH_QUERY_HISTORY = "SEARCH_QUERY_HISTORY"
+    const val DEBOUNCE_DELAY = 3000L
```
Константы лежат в пакете data, но при этому тут есть те, которые относятся к UI (вот эта задержка, нулевое время плеера, значение скругленных углов (O-o, они в коде не должны быть в таком виде, только в dimens)), их надо либо отделить в пакет UI, либо сделать отдельный пакет utils вне слоев

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/data/impl/search/HistoryRepositoryImpl.kt:
```diff
@@ -0,0 +1,37 @@
+        tempList.removeIf { it.trackId == newTrack.trackId }
+        tempList.add(0, newTrack)
+        if (tempList.size > MAX_TEMP_LIST_SIZE) {
+            tempList = tempList.subList(0, MAX_TEMP_LIST_SIZE)
```
Можно еще просто удалить последний элемент, а не создавать срез листа, он будет создан новым списком, это не эффективно. Плюс в случае с удалением элемента, саму переменную tempList можно будет сделать неизменяемой, то есть val

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/app/ThemeSwitch.kt:
```diff
@@ -0,0 +1,15 @@
+    fun switch(switchIsOn: Boolean) {
+        AppCompatDelegate.setDefaultNightMode(
+            if (switchIsOn) {
+                AppCompatDelegate.MODE_NIGHT_NO
```
У тебя по коду switchIsOn - это когда темная тема включена, а здесь перепутано, поэтому вместо светлой по дефолту загружается темная

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/data/impl/search/SearchRepositoryImpl.kt:
```diff
@@ -0,0 +1,50 @@
+    private val networkClient: NetworkClient
+) : SearchRepository {
+
+
```
Лучше делать по одной пустой строке, так не будет ощущения разорванного кода

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/data/network/RetrofitNetworkClient.kt:
```diff
@@ -0,0 +1,43 @@
+class RetrofitNetworkClient(private val context: Context) : NetworkClient {
+
+
+    private val retrofit = Retrofit.Builder()
```
И ретрофит надо будет с апи через коин внедрять

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/di/DataModule.kt:
```diff
@@ -0,0 +1,56 @@
+
+val dataModule = module {
+
+    single<SearchApi> {
```
Хм, у тебя уже даже это есть, тогда надо будет перенести апи в конструктор репозитория просто и удалить из класса

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/di/ViewModule.kt:
```diff
@@ -0,0 +1,15 @@
+        SearchViewModel(searchInteractor = get(), historyInteractor = get())
+    }
+    viewModel {
+        SettingsViewModel(settingsInteractor = get(), buttonsInteractor = get())
```
Buttons интерактор не самое говорящее название, не отображает суть. В примерах у вас ему дают название ExternalNavigator

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/domain/SearchResult.kt:
```diff
@@ -0,0 +1,11 @@
+package com.example.playlistmaker.domain
+
+//sealed class SearchResult<T> {
```
Лучше удалять закомментированный код

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/domain/api/player/PlayerInteractor.kt:
```diff
@@ -0,0 +1,22 @@
+    fun release()
+
+    fun setOnPreparedListener(listener: (() -> Unit)?)
+
```
Здесь тоже лишняя строка

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/domain/api/search/SearchRepository.kt:
```diff
@@ -0,0 +1,10 @@
+import com.example.playlistmaker.domain.SearchResult
+import com.example.playlistmaker.domain.models.search.Track
+
+
```
И тут, в следующих местах уже не будут писать

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/domain/implemantation/settings/ButtonsRepositoryImpl.kt:
```diff
@@ -0,0 +1,38 @@
+import com.example.playlistmaker.R
+import com.example.playlistmaker.domain.api.settings.ButtonsRepository
+
+class ButtonsRepositoryImpl(
```
Реализация репозитория должна быть в дата слое

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/domain/implemantation/settings/ButtonsRepositoryImpl.kt:
```diff
@@ -0,0 +1,38 @@
+class ButtonsRepositoryImpl(
+    private val context: Context
+) : ButtonsRepository {
+    override fun termsOfUse() {
```
Названия методов лучше писать с глаголами

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/domain/implemantation/settings/ButtonsRepositoryImpl.kt:
```diff
@@ -0,0 +1,38 @@
+        val text = context.getString(R.string.support_mail_text)
+        val intent = Intent(Intent.ACTION_SENDTO)
+        intent.data = Uri.parse("mailto:")
+        intent.putExtra(Intent.EXTRA_EMAIL, arrayOf(context.getString(R.string.dev_email)))
```
Чтобы каждый раз не писать intent. можно использовать метод [apply](Intent(Intent.ACTION_SENDTO)):
```
intent.apply {
    data =
    putExtra()
    и тд
}

```

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/domain/models/search/Track.kt:
```diff
@@ -0,0 +1,72 @@
+        parcel.readString(),
+        parcel.readString()
+    ) {
+    }
```
Чтобы не писать всю эту реализацию руками, можно просто использовать аннотацию [@Parcelize](https://developer.android.com/kotlin/parcelize). И тогда это все будет под капотом

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/domain/models/search/TrackResponse.kt:
```diff
@@ -0,0 +1,3 @@
+package com.example.playlistmaker.domain.models.search
+
+class TrackResponse(val results: List<Track>)
```
Этот класс нигде  не используется

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/ui/MainActivity.kt:
```diff
@@ -1,29 +1,32 @@
+        setContentView(binding.root)
 
-        binding?.buttonSearch?.setOnClickListener {
+        binding.buttonSearch.setOnClickListener {
```
Для биндинга тоже можно использовать apply

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/ui/player/AudioPlayerActivity.kt:
```diff
@@ -1,24 +1,38 @@
+            Track::class.java
+        )
+    }
+    private var player = PlayerRepositoryImpl()
```
Здесь нарушение чистой архитектуры (причем из критических замечаний).  Для этого экрана должна быть вью модель (но по идее должен быть презентер в этом спринте, а в 16 вью модель), в которую внедряется интерактор из domain, вся логика происходит во вью модели. А активити взаимодействует только с ней

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/ui/player/AudioPlayerActivity.kt:
```diff
@@ -48,6 +57,36 @@ class AudioPlayerActivity : AppCompatActivity() {
     }
 
+    private fun showTrack(track: Track) {
+        binding?.apply {
```
Зря писала рекомендацию про apply, ты ее знаешь)

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/ui/player/AudioPlayerActivity.kt:
```diff
@@ -48,6 +57,36 @@ class AudioPlayerActivity : AppCompatActivity() {
+                SimpleDateFormat("mm:ss", Locale.getDefault()).format(track.trackTimeMillis)
+            trackYearTv.text = LocalDateTime.parse(
+                track.releaseDate,
+                DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss'Z'")
```
Формат можно вынести в константу

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/ui/player/AudioPlayerActivity.kt:
```diff
@@ -48,6 +57,36 @@ class AudioPlayerActivity : AppCompatActivity() {
+                .load(track.artworkUrl512)
+                .placeholder(R.drawable.ic_placeholder)
+                .centerCrop()
+                .transform(RoundedCorners(AP_ROUNDED_CORNERS))
```
Некорректно задаются углы скругления. Для получения корректного значения закругления следует использовать следующий код: holder.itemView.resources.getDimensionPixelSize(Здесь ид ресурса, в котором указан размер радиуса в dp.). Так как значение в RoundedCorners должно передаватьсяв пикселях, а на разных устройствах разная плотность пикселей, то есть, если передавать просто число, то будет по-разному отображаться

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/ui/search/SearchViewModel.kt:
```diff
@@ -0,0 +1,123 @@
+            else -> {
+
+                val searchRunnable = Runnable { sendToServer(queryNew) }
+                handler.removeCallbacks(searchRunnable)
```
Здесь searchRunnable - это всегда новый объект, который не имеет ничего общего с добавленным в прошлый раз runnable (Вот близнецы: выглядят одинаково, но разные люди). Поэтому никакие runnable не будут удалены из хендлера. Чтобы это был всегда один объект, его надо создать один раз на уровне класса и везде использовать только его

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/ui/search/SearchViewModel.kt:
```diff
@@ -0,0 +1,123 @@
+    fun searchDebounce(queryNew: String) {
+        lastQuery = queryNew
+        when (lastQuery) {
+            queryNew.isNullOrBlank().toString() ->
```
Вот это никогда не будет работать. Смотри, у тебя queryNew - это запрос, который ввели. Когда ты делаешь queryNew.isNullOrBlank() ты получаешь булевое значение, пустой ли запрос, то есть true или false, когда ты делаешь после этого toString(), ты получаешь строку: "true" или "false". Соответственно этот код сработает только тогда, когда пользователь введет запрос "false"

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/ui/search/SearchViewModel.kt:
```diff
@@ -0,0 +1,123 @@
+        setState(State.SomeHistory(historyInteractor.read()))
+    }
+
+    fun searchDebounce(queryNew: String) {
```
Метод лучше сделать приватным, так как он используется только внутри класса

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/ui/search/SearchViewModel.kt:
```diff
@@ -0,0 +1,123 @@
+
+                val searchRunnable = Runnable { sendToServer(queryNew) }
+                handler.removeCallbacks(searchRunnable)
+                handler.postAtTime(searchRunnable, TOKEN, DEBOUNCE_DELAY)
```
Только благодаря токену runnable будут убиты в onCleared, но это также должно произойти на строку выше, как я и описала

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/ui/search/SearchViewModel.kt:
```diff
@@ -0,0 +1,123 @@
+        }
+    }
+
+    private var lastQuery: String? = null
```
По конвенции есть порядок, по корому стоит писать класс: https://kotlinlang.org/docs/coding-conventions.html#class-layout, поля лучше размещать до методов

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/ui/search/SearchViewModel.kt:
```diff
@@ -0,0 +1,123 @@
+
+    fun clearSearchHistory() {
+        historyInteractor.clear()
+        historyModification()
```
Тоже лучше в название использовать глагол, так будет понятнее

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/ui/search/SearchViewModel.kt:
```diff
@@ -0,0 +1,123 @@
+    }
+
+    fun some(queryNew: String) {
+        if (queryNew.isNullOrBlank()) {
```
на null нет смысла тут проверять, так как передается ненуллабельный String, поэтому достаточно будет проверки isBlank()

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/ui/search/SearchViewModel.kt:
```diff
@@ -0,0 +1,123 @@
+
+    fun some(queryNew: String) {
+        if (queryNew.isNullOrBlank()) {
+            Log.e("myLog", "Jopa")
```
)

nim0y replied:
Прошу прощения я не знаю как я это оставил вообще . Извини что пришлось на это обращать внимание.

Anna-Malinina replied:
Да ладно, считаем пасхалкой)

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/ui/search/SearchViewModel.kt:
```diff
@@ -0,0 +1,123 @@
+            Log.e("myLog", "Jopa")
+            setState(State.SomeHistory(historyInteractor.read()))
+        }
+        if (queryNew.isNotEmpty()) {
```
Метод isBlank() проверяет, строка пустая или состоит только из пробелов. Поэтому содержимое этого ифа можно поместить в блок else ифа выше. Потому что в else там всегда будет непустая строка (и бонусом, не состоящая только из пробелов)

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/ui/search/SearchViewModel.kt:
```diff
@@ -0,0 +1,123 @@
+        }
+    }
+
+    @SuppressLint
```
Не стоит подавлять предупреждения, тем более, что единственное предупреждение в этом методе - что его можно сделать приватным

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/ui/search/SearchViewModel.kt:
```diff
@@ -0,0 +1,123 @@
+    }
+
+    @SuppressLint
+    fun sendToServer(queryNew: String) {
```
Не совсем корректно название, на сервер ничего не отсылается здесь

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/ui/search/SearchViewModel.kt:
```diff
@@ -0,0 +1,123 @@
+                    val tempList = mutableListOf<Track>()
+                    handler.post {
+                        if (found != null) {
+                            tempList.clear()
```
Зачем очищать и так пустой список? В него при создание ничего не добавлялось

Anna-Malinina replied:
В целом все эти манипуляции со списками не нужны, так как список у нас уже есть, поэтому можно упросить:

![image](https://github.com/nim0y/PlaylistMaker/assets/129777932/c88b13ba-a43b-4961-95a0-41e8f78c28a6)

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/ui/search/TrackAdapter.kt:
```diff
@@ -0,0 +1,53 @@
+class TrackAdapter() : RecyclerView.Adapter<TrackViewHolder>() {
+
+    var tracksList: ArrayList<Track> = ArrayList()
+    lateinit var itemClickListener: ((Track) -> Unit)
```
Листенер можно передавать сразу в конструктор адаптера в виде лямбды, тогда не потребуется функциональный интерфейс, будет гораздо проще. И меньше шансов, что разработчик (ведь на проектах бывают и другие разработчики), забудет проинициализировать этот листенер извне, а так большой шанс получить ошибку, что поле не инициализировано

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/ui/search/TrackAdapter.kt:
```diff
@@ -0,0 +1,53 @@
+    var tracksList: ArrayList<Track> = ArrayList()
+    lateinit var itemClickListener: ((Track) -> Unit)
+
+    fun interface onItemClickListener {
```
Тем более, что он не используется

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/ui/search/TrackAdapter.kt:
```diff
@@ -0,0 +1,53 @@
+        return TrackViewHolder(view)
+    }
+
+    override fun getItemCount(): Int {
+        return tracksList.size
+    }
```
Более в котлин стиле будет getItemCount() = trackList.size https://kotlinlang.org/docs/coding-conventions.html#functions

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/ui/search/TrackAdapter.kt:
```diff
@@ -0,0 +1,53 @@
+}
+
+
+//private var isClickAllowed = true
```
Лучше удалить

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/ui/search/TrackViewHolder.kt:
```diff
@@ -1,19 +1,19 @@
 import java.text.SimpleDateFormat
 import java.util.Locale
 
 const val CORNER_RADIUS = 10
```
тоже самое и с этим скруглением

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/ui/search/activity/SearchActivity.kt:
```diff
@@ -0,0 +1,231 @@
+        binding?.trackRecyclerView?.adapter = searchResAdapter
+        binding?.trackRecyclerView?.itemAnimator = null
+        binding?.searchHistoryRecycleView?.adapter = searchHistoryAdapter
+        binding?.trackRecyclerView?.itemAnimator = null
```
не тот ресайклер)

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/ui/search/activity/SearchActivity.kt:
```diff
@@ -0,0 +1,231 @@
+    }
+
+    fun clearButtonVisibility(s: CharSequence?): Int {
+        return if (s.isNullOrEmpty()) {
```
можно вместо return сразу писать fun clearButtonVisibility(s: CharSequence?) = if() {} else {}

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/ui/search/activity/SearchActivity.kt:
```diff
@@ -0,0 +1,231 @@
+
+    private fun showProgressBar() {
+
+        binding?.noConnectionErrorLayout?.isVisible = false
```
здесь тоже apply был бы в тему

Anna-Malinina commented:
Беда с кликами

https://github.com/nim0y/PlaylistMaker/assets/129777932/626f201a-2aa5-4e6c-8640-f12024137173

nim0y committed:
Homework Sprint #15. Part 2
Анна,
Не передать словами испытываемое мной чувство благодарности к тебе (надеюсь можно на ты).
Не только за огромный обьём проделанной тобой работы.
Но и за то что с пониманием отнеслась к моей "слетевшей кукухе".
Постарался подчистить то что ты описывала в ревью и комментариях. Не всё получилось, но я продолжаю работу во многом благодаря тебе.
Еще раз спасибо!

nim0y committed:
Homework Sprint #15. quick fix #1
koin MediaPlayer

nim0y committed:
Homework Sprint #15. fix fix'a
Падало приложение при повторном открытии Плеера

nim0y committed:
Homework Sprint #15. fix\fix'a
Плодить коммиты больше не буду
Поторопился забыл про factory :facepalm

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/data/impl/player/PlayerRepositoryImpl.kt:
```diff
@@ -2,10 +2,12 @@ package com.example.playlistmaker.data.impl.player
+class PlayerRepositoryImpl(private val mediaPlayer: MediaPlayer) : PlayerRepository {
 
-    private var mediaPlayer = MediaPlayer()
+    //    private var mediaPlayer: MediaPlayer = MediaPlayer()
```
Удалить можно насовсем)

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/ui/player/AudioPlayerActivity.kt:
```diff
@@ -2,55 +2,48 @@ package com.example.playlistmaker.ui.player
-    private var playerTime: TextView? = null
-
-    private val mainThreadHandler = Handler(Looper.getMainLooper())
+    private lateinit var track: Track
```
К 16 спринту надо будет убрать трек из активити во вью модель, там будет пункт, что активити должна получать все необходимые для отображения данные из вью модели

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/ui/player/AudioPlayerActivity.kt:
```diff
@@ -70,83 +63,43 @@ class AudioPlayerActivity : AppCompatActivity() {
                 .placeholder(R.drawable.ic_placeholder)
                 .centerCrop()
-                .transform(RoundedCorners(AP_ROUNDED_CORNERS))
+                .transform(RoundedCorners(applicationContext.resources.getDimensionPixelSize(R.dimen.track_corner_radious)))
```
К ресурсам есть доступ напрямую из активити, можно не вызывать контекст приложения

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/ui/player/AudioPlayerActivity.kt:
```diff
@@ -70,83 +63,43 @@ class AudioPlayerActivity : AppCompatActivity() {
+            PlayerState.PLAYING_STATE -> setIconPause()
+            PlayerState.PAUSE_STATE -> setIconPlay()
+            PlayerState.PREPARATION_STATE -> setIconPlay()
+            PlayerState.DEFAULT_STATE -> {}
```
Еще можно писать `-> Unit` (Это кому как нравится)

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/ui/player/AudioPlayerViewModel.kt:
```diff
@@ -0,0 +1,84 @@
+    }
+
+    fun playControl() {
+        when (playerInteractor.playerCheck()) {
```
Здесь хватит if, так как всего два состояния: true или false. Так рекомендуется: https://kotlinlang.org/docs/coding-conventions.html#if-versus-when)

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/ui/search/SearchViewModel.kt:
```diff
@@ -24,100 +22,87 @@ class SearchViewModel(
+        searchRunnable = Runnable { request() }
+        handler.removeCallbacksAndMessages(TOKEN)
+        val time = SystemClock.uptimeMillis() + DEBOUNCE_DELAY
+        handler.postAtTime(searchRunnable!!, TOKEN, time)
```
Можно не использовать точное время для запуска, достаточно будет postDelayed

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/ui/search/SearchViewModel.kt:
```diff
@@ -24,100 +22,87 @@ class SearchViewModel(
+    }
+
+    fun onDestroyHandlerRemove() {
+        handler.removeCallbacksAndMessages(Any())
```
Any() здесь будет совершенно другим объектом, отличным от токена (так как создается новый, с новой ссылкой). Поэтому для удаления надо использовать твой токен, либо null, если мы хотим удалить все, что могли добавить в хендлер

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/ui/search/SearchViewModel.kt:
```diff
@@ -24,100 +22,87 @@ class SearchViewModel(
 
-    private var lastQuery: String? = null
+    private fun searchDebounce(request: () -> Unit) {
+        searchRunnable = Runnable { request() }
```
При каждом вызове метода поиска searchRunnable перезаписывается новым объектом, то есть мы теряем предыдущий, и, если нам понадобится удалить предыдущий через searchRunnable, то сделать этого мы уже не сможем. searchRunnable надо инициализировать сразу при объявление в начале класса (решить вопрос с параметром можно, если запрос сохранять в методе queryDebounce в приватную переменную класса, и использовать ее внутри searchRequest)

nim0y replied:
Если ты меня слышишь, знаю что это не твоя обязанность мне обьяснять. 
Но попробую уточнить у тебя . В любом случае , не устану тебя благодарить ;) Спасибо!
 Как более логично выглядит этот блок ? Или я в бутылку лезу уже?
 
Var#1
```
private var searchRunnable: Runnable? = Runnable {
        val queryNew = lastQuery?:""
        searchRequest(queryNew) }
fun queryDebounce(queryNew: String?) {
        lastQuery = queryNew
        if (queryNew.isNullOrEmpty()) {
            setState(State.SomeHistory(historyInteractor.read()))
        } else {
            searchDebounce (queryNew)
        }
    }

    private fun searchDebounce(changedText:String) {
        this.lastQuery = changedText
        handler.removeCallbacksAndMessages(TOKEN)
        handler.postDelayed(searchRunnable!!, TOKEN, DEBOUNCE_DELAY)

    }
```
var#2
```
private var searchRunnable: Runnable? = Runnable { request?.invoke() }
    private var lastQuery: String? = null
    private var request: (() -> Unit)? = null
    
    fun queryDebounce(queryNew: String?) {
        lastQuery = queryNew
        if (queryNew.isNullOrEmpty()) {
            setState(State.SomeHistory(historyInteractor.read()))
        } else {
            searchDebounce { searchRequest(queryNew) }
        }
    }

    private fun searchDebounce(request: () -> Unit) {
        this.request = request
        handler.removeCallbacksAndMessages(TOKEN)
        handler.postDelayed(searchRunnable!!, TOKEN, DEBOUNCE_DELAY)

    }
```
```

Anna-Malinina replied:
Привет! Я в отпуске до 21 числа, ноутбука рядом нет( если не возражаешь, я отвечу 21-22 числа)

nim0y replied:
Конечно, извини за беспокойство.

Anna-Malinina replied:
Ничего страшного, я буду рада помочь)

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/ui/search/SearchViewModel.kt:
```diff
@@ -24,100 +22,87 @@ class SearchViewModel(
 
     companion object {
-        private val TOKEN = Any()
+        val TOKEN = Any()
```
токену лучше вернуть приватность, он же не используется нигде извне

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/ui/search/TrackAdapter.kt:
```diff
@@ -27,27 +23,8 @@ class TrackAdapter() : RecyclerView.Adapter<TrackViewHolder>() {
         holder.bind(tracksList[position])
         holder.itemView.setOnClickListener {
-            itemClickListener?.invoke(tracksList[position])
+            itemClickListener.invoke(tracksList[position])
```
можно вызывать сразу через круглые скобки itemClickListener(), а назвать его onItemClick, например, так как теперь это лямбда, а не объект листенера

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/ui/settings/SettingsViewModel.kt:
```diff
@@ -6,12 +6,12 @@ import androidx.lifecycle.MutableLiveData
+    private val externalNavigatorInteractor: ExternalNavigatorInteractor
 ) : ViewModel() {
 
     private val switchOnLiveData = MutableLiveData<Boolean>()
```
Причина бага с переключателем на 25 строке:

![image](https://github.com/nim0y/PlaylistMaker/assets/129777932/0e336cd9-2bca-4426-bf70-46a02f8a61dd)

Вот здесь у тебя сбрасывается состояние, так как ставишь false, а не darkTheme (а еще признаки лучше писать с глаголом to be, то есть isDarkThemeEnabled или что-то такое. Чтобы сразу стало понятно, что будет только true или false

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/ui/search/activity/SearchActivity.kt:
Сейчас разберем, куда копать с багами поиска

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/ui/search/activity/SearchActivity.kt:
```diff
@@ -42,20 +40,22 @@ class SearchActivity : AppCompatActivity() {
-            openAudioPlayer(track = it)
+            if (clickDebounce()) {
+                vm.addTrackToHistory(it)
+                vm.queryDebounce(binding?.editQuery?.text?.toString() ?: "")
```
Вот здесь поиск не требуется, то есть, когда мы нажимаем на открытие трека, не должен происходить поиск треков опять

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/ui/search/activity/SearchActivity.kt:
```diff
@@ -69,45 +69,46 @@ class SearchActivity : AppCompatActivity() {
         }
         binding?.searchRefreshButton?.setOnClickListener {
-            vm.searchDebounce(binding?.editQuery?.text?.toString() ?: "")
+            vm.queryDebounce(binding?.editQuery?.text?.toString() ?: "")
```
По идее, если пользователь нажал на обновить, то ему точно не захочется ждать, пока пройдут три секунды, надо начинать поиск без откладывания

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/ui/search/activity/SearchActivity.kt:
```diff
@@ -69,45 +69,46 @@ class SearchActivity : AppCompatActivity() {
-        })
+
+        }
+        textWatcher.let {
```
В целом тут let не нужен, так как можно сразу написать binding?.editQuery?.addTextChangedListener(textWatcher), на две строки меньше кода)

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/ui/search/SearchViewModel.kt:
```diff
@@ -24,100 +22,87 @@ class SearchViewModel(
 
     init {
-        Log.e("myLog", "------${_searchState.value}")
         setState(State.SomeHistory(historyInteractor.read()))
```
По ТЗ история должна показываться только тогда, когда поле ввода в фокусе. То есть если зашел на страницу поиска, то она должна быть пустой. Затем, если нажал на поле ввода и история не пустая, то история отображается. Это несоответствие функционалу у тебя, может не пройти в 16 спринте, если ревьюер будет такой же дотошный как я

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/ui/search/activity/SearchActivity.kt:
```diff
@@ -42,20 +40,22 @@ class SearchActivity : AppCompatActivity() {
-            vm.searchDebounce(binding?.editQuery?.text?.toString() ?: "")
-            openAudioPlayer(track = it)
+            if (clickDebounce()) {
+                vm.addTrackToHistory(it)
```
Добавление трека в историю не должно означать то, что эту историю надо показать. Это тоже причина бага с поиском. Корректное поведение: пользователь нажал на трек, трек ему открылся, он вернулся на экран поиска, а там все еще результат поиска, а не история.  Нужно доработать метод во вью модели, чтобы она не отображала историю каждый раз после нажатия на трек

Anna-Malinina commented on app/src/main/java/com/example/playlistmaker/ui/search/SearchViewModel.kt:
```diff
@@ -24,100 +22,87 @@ class SearchViewModel(
+        if (queryNew.isNullOrEmpty()) {
+            setState(State.SomeHistory(historyInteractor.read()))
+        } else {
+            setState(State.Load)
```
Нужно перенести это состояние в то место, где реально начинается загрузка. Пользователь должен увидеть прогресс бар только после задержки, а не на ввод каждого символа. (То есть внутри runnable)

nim0y committed:
MVVM

nim0y committed:
MVVM
